Documentation/design-docs/host-tracing.md

```diff
@@ -0,0 +1,74 @@
+
+## Components involved
+Host tracing spans various components:
+* `dotnet` or `apphost` executable
+* `hostfxr` library
+* `hostpolicy` library
+
+If a custom host is used it will typically replace the executables and possible even some of the libraries. The custom host can use either `hostfx` or `hostpolicy` directly. So each of these components must support tracing on its own along with the configuration. In case several of these components are used together, they must cooperate and "share" the tracing setup.
+
+Currently (as of .NET Core 2.1) all these components include the same code for tracing but act completely independently. VIa the fact that they share the same code, the end behavior is that they all write their trace output to the same location `stderr`. The only synchronization currently implemented is that each component fully flushes all internal tracing buffers before transferring control to a different component.
+
+## New trace routing
+By default tracing is disabled and is not routed anywhere. If it's enabled it is routed through an instance of `host_trace_listener` interface.
+``` C++
+struct host_trace_listener
+{
+    void trace_verbose(const pal::char_t *message);
+    void trace_info(const pal::char_t *message);
+    void trace_warning(const pal::char_t *message);
+    void trace_error(const pal::char_t *message);
+    void flush();
+}
+```
+
+The `message` parameter is a standard `NUL` terminated string.
+
+The host components provide two built in implementations of this interface:
+* `stderr_trace_listener` which implements the .NET Core 2.1 behavior of routing tracing to the `stderr` output of the process. This is the default listener which is activated by setting `COREHOST_TRACE=1` environment variable.
+* `file_trace_listener` which routes tracing into a file. This listener is activated by setting the `COREHOST_TRACE=1` and also providing the full path to the trace file in `COREHOST_TRACEFILE=<path>` environment variable.
+
+Custom host can provide its own implementation of `host_trace_listener` by calling:
+``` C++
+void hostfxr_set_trace_listener(host_trace_listener* listener)
+```
+from the `hostfxr` component or
+``` C++
+void corehost_set_trace_listener(host_trace_listener* listener)
+```
+from the `hostpolicy` component.
+
+The functions behave exactly the same in both components. The `listener` parameter can be:
+* a pointer to an implementation of `host_trace_listener` which is then registered the only listener for all tracing.
+* `NULL` value which unregisters any previously registered listener. After this call tracing is disabled.
+
+Custom host can and should register the trace listener as the first thing it does with the respective host component to ensure that all tracing is routed to it.  
+
+Only one trace listener can be registered at any given time.  
```

Comment:
So if there is a trace listener, then no output to file and not output to stderr. Correct?

A potential concern down the road is that the host tests currently use tracing to verify behavior, so it a test wishes to verify a custom ASPNET host (for example) it won't get the stderr messages since ASPNET will likely have it's own listener.

Reply:
I think the model here is that we can have multiple listeners based or cooperative protocol in this apps. So, we'll need to run the tests in cooperation with ASP.net.

Reply:
This is the one thing I'm still a bit unsure myself. We could keep the current behavior completely as is and add the custom host registered listeners on top. (So if the env variable is set we would route the tracing to both stderr and the listener). Custom hosts could choose to disable this by explicitly resetting the env variable on early on.
If this sounds good I'll make the change.

Reply:
I changed this in the doc to allow both `stderr` and custom listener at the same time.